doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+:title: Users Guide: Promotion
+
+.. _ug-promotion:
+
+Promotion: implicit data parallelism
+====================================
+
+In Chapel, *function promotion* (or simply *promotion*) refers to an
+implicit form of data parallelism that is triggered by passing a
+collection of values to an argument that expects a single value.  More
+precisely, when an array of type *t* (or a range or domain whose index
+type is *t*) is passed to a function argument of type *t*, that
```

Comment:
Oh, this was a misunderstanding on my part.  I knew that we didn't support promotion by arbitrary iterable expressions at present.  For example, if you define a class with all the normal these() iterators, we don't permit that class to promote functions... (note that I think this is what we ultimately want... it's just not what we did in the early days when we didn't yet have a parallel iterator story in place).

Instead, the current Chapel source code trigger for something to enable promotion is the presence of a member field called 'promotionType'.  This mechanism was not intended for end-users and is currently only used by ranges, domains, and arrays -- so I thought I was being precise in the above (even though that's not where we ultimately want to be).  What I failed to account for in my 'git grep' was that the compiler propagates this information for iteration expressions like the one you note... oops.

It's pretty fragile in when it does or does not do the propagation, though, so I don't know how broadly we should take the statement above at present.  I could try to be precise (which I think is "ranges, domains, arrays, or forall expressions over these types"), or I could stick with the present wording to be conservative and consider the lack of anonymous forall expressions a sin of omission, or I could say "all iterable expressions" and let someone call it a (big) bug when they run afoul of the current limitation (really more of an unimplemented feature that we've never formally discussed).

On the day that we do support promotion by iteration, I think the approach you suggest is definitely right.  I've been trying to write the users guide more to describe what we support than to what we hope to support in the future which is why I took this approach (and would be open to trying to describe what we have today more precisely if that's what you think is best... I'm just not entirely sure where the line between works and doesn't falls).